### PR TITLE
Implement task-based array shuffle

### PR DESCRIPTION
**Pull request by phofl**
- [ ] xref #11257
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

This is a first rough implementation of an array-based shuffle that uses the task implementation. Not polished and probably has a few edge cases that aren't covered yet.

TLDR: We are preserving chunk sizes (roughly) compared to take but we still have a very large graph, which strains the scheduler pretty heavily. Subsequent results are easier to process though in case of a random access pattern because the chunks stay reasonable. An improvement over random access take generally speaking, but still not very good

Let's mess around with a toy example:

```
import dask.array as da

n = 1500
arr = da.random.random((n, n, n), chunks=
(50, 50, 50))
arr
```

Screenshot 2024-07-31 at 11 57 00

Our indexer has a random access pattern (and we access 5 elements at a time, but we could do random stuff here).

```
import numpy as np
from dask.core import flatten

idx = np.arange(0, n)
np.random.shuffle(idx)

indexer = []
for i in range(0, n, 5):
    indexer.append(idx[i:i+5].tolist())
take_indexer = list(flatten(indexer))

indexer
```

```
result = arr.shuffle(indexer, axis=1)
result
```

Screenshot 2024-07-31 at 11 57 21

Shuffling keeps the number of chunks consistent. Doing the equivalent take blows up the number of chunks

```
expected = arr[slice(None), take_indexer, slice(None)]
expected
```

Screenshot 2024-07-31 at 11 57 35

The number of tasks are in the same ballpark though, the growth is still exponential in a case of an all to all pattern.

 
Screenshot 2024-07-31 at 11 57 54

Generating close to a million tasks for a 27k chunks array is not great.

### CODE REVIEW

**Comment by fjetter** [comment]
I assume this is pretty much equivalent to `arr.take(...).rechunk(arr.chunks)` isn't it?

**Comment by phofl** [comment]
> I assume this is pretty much equivalent to arr.take(...).rechunk(arr.chunks) isn't it?

roughly, we can be smarter with respect to how many tasks we create, i.e. if you have a 6 x 6 array with 3 x 3 chunks and an indexer on axis 1 of

``[[0, 3, 1], [4, 2, 5]]``

take would create a task per element, because they aren't in order. We can sort and then re-order to only slice once per chunk. Shuffle also ensures that a group of elements ends up in the same chunk.

That said, worst case task graph size is roughly the same yes.

**Comment by mrocklin** [comment]
A couple questions / thoughts (but I haven't thought about this as much, so feel free to disregard)

### Multi-stage

Let's imagine a pathological case where we have two chunks and and indexer that switches back and forth between them

```
chunks = ((100000, 100000),)
indexer = [0, 100000, 1, 100001, 2, 100002]
```

Optimally we should be able to do this in around 6 tasks I think, assuming we're willing to do the take in a couple of stages, in worst case we'll do it in 200,000 tasks.  What happens with this implementation?

Also, it may not make sense to invest in the multi-stage implementation here for various reasons, I'm just making sure that we have that option on our radar.

### Lots of Rechunking 

If we did the multi-stage thing, then suddenly we're strongly incentivized to rechunk ahead of time to make the take dimension very large.  This could be useful for two reasons:

1.  The input chunk sizes are too small (I think that we should be willing to override small input chunk sizes for intermediate results)
2. The input chunking direction isn't optimal for the take operation


Just some thoughts.  It could be that we've already been thinking about these or that they're bad for some reason.  No need to anchor on them, just bringing them up in case they're useful.

**Comment by mrocklin** [comment]
> We do this in 6 tasks at the moment. We sort the indexer and then resort after selecting the data

Oh grand!

So I guess we're experiencing here a task blowup of shape/chunkshape for the take dimension, in this case that's 1500/50 == 30x.  So 27k * 30 == ~800k, which tracks.  This is a lot better than the number of chunk transitions in the indexer, which is what we had before.  This seems great!

**Comment by dcherian** [comment]
Looks like the approach in https://github.com/dask/dask/pull/3901?

**Comment by phofl** [comment]
Yep very similar approach. Task graph will be smaller but both leverage the same technique.

This is not the state that I would like to end up in though.


A more general question: Is this the kind of API could be used in Xarray for a group by-transform or group by-reduce if you don't want to do a tree reduction? Basically what we discussed here: https://github.com/pydata/xarray/issues/9220

**Comment by phofl** [comment]
This should be ready for a closer look.

@dcherian is this an API that would satisfy the XArray requirements for a groupby - transform approach?

We can add the P2P backend in the background after we align on how we want to process things

**Comment by phofl** [comment]
With #11267, take performance is a lot better:

https://github.com/coiled/benchmarks/actions/runs/10202090066/artifacts/1765715083

```
def _create_indexer(n, chunk_n):
    idx = np.arange(0, n)
    np.random.shuffle(idx)

    indexer = []
    for i in range(0, n, chunk_n):
        indexer.append(idx[i : i + chunk_n].tolist())
    return indexer


def test_take(small_client, new_array):
    n = 2000
    chunk_n = 250
    x = da.random.random((n, n, n), chunks=(chunk_n, chunk_n, chunk_n))
    indexer = list(flatten(_create_indexer(n, chunk_n)))
    x[:, indexer, :].sum().compute()
```

This is the largest array that still works with the current take implementation, everything larger just kills the cluster

File: dask/array/_shuffle.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    number of chunks consistent or at least manageable.
+
+    Parameters
+    ----------
+    x: dask array
+        Array to be shuffled.
+    indexer:  list[list[int]]
+        The indexer that determins which elements along the dimension will end up in the
+        same chunk. Multiple groups can be in the same chunk to avoid fragmentation, but
+        each group will end up in exactly one chunk.
```
Comment (dcherian) [inline dask/array/_shuffle.py:32]: This behaviour is problematic for

```python
import xarray as xr
import dask.array

time = xr.date_range("2001-01-01", "2010-01-01", freq="D")
ds = xr.Dataset(
    {"data": (("x", "y", "time"), dask.array.random.random((10000, 1000, len(time)), chunks=(-1, -1, 1)))},
    coords={"time": ("time", time)},
)
ds.resample(time="Y")._group_indices
```

Comment (phofl) [reply dask/array/_shuffle.py:32]: Can you say a bit more?

Is the problem that the chunks will become too large or that we fuse multiple groups together?

If the latter, then we could add control for this behavior to turn fusion of groups off.

If chunks that are too large:
Would your preferred alternative be to have one group in multiple chunks along the time dimension in your example or split the other dimensions into multiple chunks instead of -1 and -1?

Comment (dcherian) [reply dask/array/_shuffle.py:32]: > one group in multiple chunks along the time dimension in your

I feel like its preferrable to do this and then have the user do the rechunk-to-one-block-while-adjusting-others.

Comment (phofl) [reply dask/array/_shuffle.py:32]: Is the user in this case xarray or users from xarray?

When we put one group into multiple chunks, all chunks that have parts of the group should be in consecutive chunks, so that 

x = x.rechunk(("x": "foo", "y": "bar", "time": 365})

would put all values of said group into a single chunk in your example?

File: dask/array/_shuffle.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    The algorithm tries to balance the chunksizes as much as possible to ideally keep the
+    number of chunks consistent or at least manageable.
+
+    Parameters
+    ----------
+    x: dask array
+        Array to be shuffled.
+    indexer:  list[list[int]]
+        The indexer that determins which elements along the dimension will end up in the
```
Comment (hendrikmakait) [inline dask/array/_shuffle.py:30]:
```suggestion
        The indexer that determines which elements along the dimension will end up in the
```

File: dask/array/core.py [hunk 3]
```diff
@@ -2762,6 +2762,23 @@ def rechunk(
 
         return rechunk(self, chunks, threshold, block_size_limit, balance, method)
 
+    def shuffle(
+        self,
+        indexer: list[list[int]],
+        axis,
+    ):
+        """Convert blocks in dask array x for new chunks.
```
Comment (hendrikmakait) [inline dask/array/core.py:2770]: I don't understand this docstring.

File: dask/array/tests/test_shuffle.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    assert result.chunks[0] == darr.chunks[0]
+    assert result.chunks[1] == chunks
+
+
+def test_shuffle_config_tolerance(arr, darr):
+    indexer = [[1, 5, 6], [0, 3], [4, 2, 7]]
+    chunks = (3, 2, 3)
+    with dask.config.set({"array.shuffle.chunksize-tolerance": 1}):
+        result = darr.shuffle(indexer, axis=1)
+    expected = arr[:, list(flatten(indexer))]
```
Comment (hendrikmakait) [inline dask/array/tests/test_shuffle.py:44]: nit: I'd personally find this test more readable if we just replaced `list(flatten(indexer))` with the corresponding literal values.

File: dask/array/tests/test_shuffle.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+)
+def test_shuffle(arr, darr, indexer, chunks):
+    result = darr.shuffle(indexer, axis=1)
+    expected = arr[:, list(flatten(indexer))]
+    assert_eq(result, expected)
+    assert result.chunks[0] == darr.chunks[0]
+    assert result.chunks[1] == chunks
+
+
+def test_shuffle_config_tolerance(arr, darr):
```
Comment (hendrikmakait) [inline dask/array/tests/test_shuffle.py:39]: Should we test different tolerances?

File: dask/dask-schema.yaml [hunk 6]
```diff
@@ -135,6 +135,17 @@ properties:
...
+      shuffle:
+        type: object
+        properties:
+
+          chunksize-tolerance:
+            type: number
+            description: |
+              Tolerance for the shuffle algorithm when creating output chunks.
+              Default is 1.25. This means that the shuffle algorithm can exceed
+              the average input chunk size along this dimension by 25%.
```
Comment (hendrikmakait) [inline dask/dask-schema.yaml:147]: Do we also impose a lower limit? When thinking about tolerances, I think about limits in both directions. If it's only an upper limit, it might make sense to make this explicit in the naming.

File: dask/array/_shuffle.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    might contain multiple groups to avoid fragmentation of the array.
+
+    The algorithm tries to balance the chunksizes as much as possible to ideally keep the
+    number of chunks consistent or at least manageable.
+
+    Parameters
+    ----------
+    x: dask array
+        Array to be shuffled.
+    indexer:  list[list[int]]
```
Comment (hendrikmakait) [inline dask/array/_shuffle.py:29]: nit: It might be safer to convert the indexer to a `tuple[tuple[int]]` internally to avoid unintentional modification.

File: dask/array/_shuffle.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            f"Axis {axis} is out of bounds for array with {len(chunks)} axes"
+        )
+
+    if max(map(max, indexer)) >= sum(chunks[axis]):
+        raise IndexError(
+            f"Indexer contains out of bounds index. Dimension only has {sum(chunks[axis])} elements."
+        )
+
+    chunksize_tolerance = config.get("array.shuffle.chunksize-tolerance")
+    average_chunk_size = int(
```
Comment (hendrikmakait) [inline dask/array/_shuffle.py:92]: nit: Rename since this is the average multiplied by the tolerance.

```suggestion
    chunk_size_limit = int(
```

File: dask/array/_shuffle.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    for idx in indexer:
+        if (
+            len(current_chunk) + len(idx) > average_chunk_size
+            and len(current_chunk) > 0
+        ):
+            new_chunks.append(current_chunk)
+            current_chunk = idx.copy()
+        else:
+            current_chunk.extend(idx)
+            if len(current_chunk) > average_chunk_size / 1.25:
```
Comment (hendrikmakait) [inline dask/array/_shuffle.py:107]: What's the purpose of the constant `1.25`?

File: dask/array/_shuffle.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            for c, b_start, b_end in zip(
+                source_chunk_nr, taker_boundary[:-1], taker_boundary[1:]
+            ):
+                # insert our axis chunk id into the chunk_tuple
+                chunk_key = convert_key(chunk_tuple, c, axis)
+                name = (split_name, next(split_name_suffixes))
+                this_slice = slices.copy()
+
+                # Cache the takers to allow de-duplication when serializing
+                # Ugly!
```
Comment (hendrikmakait) [inline dask/array/_shuffle.py:154]:
```suggestion
```

Comment (phofl) [reply dask/array/_shuffle.py:154]: Hm I wanted to express that this is necessary but very ugly, so this is kind of intended

File: dask/array/_shuffle.py (hunk 9 above)
Comment (phofl) [reply dask/array/_shuffle.py:107]: oh, needs to be tolerance

File: dask/dask-schema.yaml (hunk 6 above)
Comment (phofl) [reply dask/dask-schema.yaml:147]: only upper,

File: dask/array/tests/test_shuffle.py (hunk 5 above)
Comment (phofl) [reply dask/array/tests/test_shuffle.py:39]: sure

File: dask/array/core.py (hunk 3 above)
Comment (phofl) [reply dask/array/core.py:2770]: probably some kind of copy paste error

File: dask/array/tests/test_shuffle.py (hunk 4 above)
Comment (phofl) [reply dask/array/tests/test_shuffle.py:44]: sounds good

File: dask/array/_shuffle.py (hunk 7 above)
Comment (phofl) [reply dask/array/_shuffle.py:29]: I am messing with list methods later, do you want to protect against modifying the input?

Comment (hendrikmakait) [reply dask/array/_shuffle.py:29]: > do you want to protect against modifying the input?

Yes, I was wondering if that would be helpful here.

Comment (phofl) [reply dask/array/_shuffle.py:29]: I would rather copy in this case, you ok with?

Comment (hendrikmakait) [reply dask/array/_shuffle.py:30]: sounds good!

File: dask/array/_shuffle.py (hunk 10 above)
Comment (hendrikmakait) [reply dask/array/_shuffle.py:151]: Ah, makes sense.

**Review by hendrikmakait — APPROVED** [review]
LGTM assuming CI is green. Thanks, @phofl

**Comment by dcherian** [comment]
> s this an API that would satisfy the XArray requirements for a groupby - transform approach?

I think this gets part of the way there, yes. https://github.com/pydata/xarray/issues/9220#issuecomment-2269310682

File: dask/array/_shuffle.py (hunk 1 above)
Comment (dcherian) [reply dask/array/_shuffle.py:33]: ^ yes. Once the groups are in order, we can use slicing to extract each group and pass to the user.

Basically it'd be nice to preserve the cases where the map approach works nicely today.

```
array.resample(time="Y").reduce(np.mean)
```

Today this dispatches to `dask.array.mean` with each group split across multiple chunks, so we get the tree reduction.

Comment (phofl) [reply dask/array/_shuffle.py:33]: Yep ok, got it. This is what you would use take for after https://github.com/dask/dask/pull/11267 (what we discussed in the other issue). take uses the same algorithm under the hood but is not placing everything into a single group.
